# A scientific "Hello, World!" program

Traditionally, the first program one creates when learning a new programming language is the so-called _"Hello, World!"_ program, which simply writes a message to the console. In Python, this task is very easy. We use the `print` command as shown below:


In [1]:
print("Hello, World!")

Hello, World!


The command `print()` is a built-in function that writes the output to the screen, or in this case, directly below the code cell.

There are a few important things to note in the code already:

- Because `print` is a _function_, we need to use parentheses when we call it.
- We put the function _argument_ inside the parentheses. Functions can take zero, one, or multiple arguments.
- In this case, the argument is a text string, which we have to put inside either single or double-quotes (`'` or `"`). These are functionally equivalent, but you need to use the same character on both ends.

The "Hello, World!" program isn't terribly exciting, so let us instead do a more scientific variant where we compute some simple expression and write out the results. As a recurring example, we will use the following mathematical formula for kinematics:

$$
s(t) = v_0 t + \frac{1}{2}at^2
$$

Here, $s$ is the distance traveled by an object after time $t$ with a constant acceleration $a$, starting with an initial velocity $v_0$.

As an example, let us compute the distance traveled for $t = 0.5$, $v_0 = 2$, and $a = 0.2$.


In [2]:
t = 0.5
v0 = 2
a = 0.2

s = v0 * t + 0.5 * a * t**2

Here we first assign the three variables `t`, `v0`, and `a`. Note the following:

- We use the `=` symbol to define a variable.
- The left-hand side is the name of the variable, while the right-hand side is its value.
- Technically speaking, the assignment operation creates an _object_ in memory, and the left-hand side becomes a reference name for that object.

For the last line, we are also defining a variable by writing `s = ...`. But in this case, we do not simply hard-code the value; instead, we define it as the result of a mathematical computation. How is this interpreted by Python?

- Any time a variable is assigned, the right-hand side is first _evaluated_, and the resulting object is what is stored as the new variable.

So Python computes the right-hand side and stores the result in our `s` variable. But how do we figure out what the result actually is? We `print` it!


In [3]:
print(s)

1.025


In this case, we are not printing a text string, but a variable. Python then prints the _value_ of the variable, which is `1.025`.

Note that `s` only stores the _result_ of the evaluation. It does not remember "where it came from." What we mean by that is that the formula itself is in no way remembered by Python. If you reassign `t`, for example, the value of `s` does _not_ change. You will need to physically recompute `s` if you want to update it.

<br>

**Your turn!**

Repeat this example in the cell below, but with the values $t = 1.2$, $v_0 = 4$, and $a = -0.1$.


In [ ]:
# Write your code here...

## Python as a calculator

Our scientific "Hello, World!" example contained a simple computation, and there are a few things worth noting about how Python does math:

- Standard arithmetic is written using the symbols: `+`, `-`, `/`, and `*`.
- Python has no _implicit_ multiplication. You must explicitly write `*` (e.g., `a * t`, not `at`).
- Exponents use a double asterisk: `**` (e.g., $t^2$ is `t**2`).
- Python performs mathematical operations following standard algebraic order of operations (PEMDAS).

This last point is highly useful because it means we can simply write out formulas as we would by hand, and Python will compute them correctly. You can, of course, add parentheses to force a specific order of operations. Other built-in mathematical operators that are useful include `//` for integer division, `%` for modulo (remainder), and `abs(x)` for absolute value.

_Tip: Adding whitespace around variables and operators when writing mathematical expressions does not affect the evaluation whatsoever. Feel free to add spaces in your equations to give them some room to breathe, which makes your code much more readable! For instance, we highly recommend consistently putting a space before and after the assignment operator (`=`)._

### Note on integer division

Note that dividing integers in Python 3 automatically gives us a decimal (or a `float` object, to be precise). This is different from older programming languages like C/C++ and also different from Python 2. If you specifically want _integer division_ (where the decimal is dropped), you need to use `//`.


In [4]:
print(2 / 5)
print(12 // 5)

0.4
2


## Importing additional mathematical functions

The built-in operations only cover the most fundamental arithmetic. If we want more complex operations, such as roots, exponentials, or trigonometry, we need to _import_ this functionality. The standard Python package `math` contains all common mathematical functions.

To import a specific function from a given package, you can write it as follows:


In [5]:
from math import sqrt

print(sqrt(81))

9.0


Or you can instead import the whole package and call functions from inside it:


In [6]:
import math

print(math.sqrt(81))

9.0


There is a third option, which is to import _everything_ from a given package at once. This is done by using an asterisk (`*`):

```python
from math import *
```

This is known as a _wildcard_ import. In general, wildcard imports are highly discouraged in professional programming because it becomes unclear which functions come from which packages, causing "namespace pollution." However, it is not uncommon to see it used in quick tutorials, so it is useful to know what it does.

Here are some of the most common functions contained in the `math` package:

| Function               | Mathematical notation | Python         |
| ---------------------- | --------------------- | -------------- |
| Square root            | $\sqrt{x}$            | `sqrt(x)`      |
| Exponential function   | $e^x$                 | `exp(x)`       |
| Natural logarithm      | $\ln x$               | `log(x)`       |
| Base-10 logarithm      | $\log_{10} x$         | `log10(x)`     |
| Base-2 logarithm       | $\log_2 x$            | `log2(x)`      |
| General base logarithm | $\log_b x$            | `log(x, b)`    |
| Sine                   | $\sin x$              | `sin(x)`       |
| Inverse Sine           | $\arcsin x$           | `asin(x)`      |
| Hyperbolic Sine        | $\sinh x$             | `sinh(x)`      |
| Factorial              | $n!$                  | `factorial(n)` |

A few things to note:

- All functions in Python are called with parentheses surrounding the argument, even if mathematical notation sometimes omits them.
- Trigonometric functions compute in _radians_, not degrees.
- This list is not comprehensive. For a full overview, check the [official reference](https://docs.python.org/3/library/math.html).

In addition to these functions, the `math` package also contains common constants, such as $\pi$ and $e$.


In [7]:
from math import pi, e

print(pi)
print(e)

3.141592653589793
2.718281828459045


<br>

**Your turn!**

Using Python as a calculator and the `math` package, compute the following expressions:

1. $\frac{4\pi}{3}$
2. $12!$
3. $e^4 \cos\left(\frac{3\pi}{2}\right)$


In [ ]:
# Your code here...

## Scientific notation

One final note on using Python for computations is that Python natively supports scientific notation, which is incredibly helpful when defining very large or very small numbers.


In [9]:
print(2.4e3)
print(3.3e-4)

2400.0
0.00033


Here, `2.4e3` means $2.4 \times 10^3$, and `3.3e-4` means $3.3 \times 10^{-4}$.

Note that this same notation is often used by Python when outputting results:


In [10]:
print(6.4e2 * 5.2e-9)

3.328e-06


When doing physiological modeling, it is critical to keep track of and be consistent in your choice of units. Using compact scientific notation is an excellent way to keep your code neat and readable!
